src/coreComponents/physicsSolvers/fluidFlow/wells/PerforationFluxKernels.hpp

```diff
@@ -0,0 +1,890 @@
+                          bool const disableReservoirToWellFlow ):
+    m_resPres( compFlowAccessors.get( fields::flow::pressure {} )),
+    m_resPhaseVolFrac( compFlowAccessors.get( fields::flow::phaseVolumeFraction {} )),
+    m_dResPhaseVolFrac( compFlowAccessors.get( fields::flow::dPhaseVolumeFraction {} )),
+    m_dResCompFrac_dCompDens( compFlowAccessors.get( fields::flow::dGlobalCompFraction_dGlobalCompDensity {} )),
+    m_resPhaseDens( multiFluidAccessors.get( fields::multifluid::phaseDensity {} )),
+    m_dResPhaseDens( multiFluidAccessors.get( fields::multifluid::dPhaseDensity {} )),
+    m_resPhaseVisc( multiFluidAccessors.get( fields::multifluid::phaseViscosity {} )),
+    m_dResPhaseVisc( multiFluidAccessors.get( fields::multifluid::dPhaseViscosity {} )),
+    m_resPhaseCompFrac( multiFluidAccessors.get( fields::multifluid::phaseCompFraction {} )),
+    m_dResPhaseCompFrac( multiFluidAccessors.get( fields::multifluid::dPhaseCompFraction {} )),
+    m_resPhaseRelPerm( relPermAccessors.get( fields::relperm::phaseRelPerm {} )),
+    m_dResPhaseRelPerm_dPhaseVolFrac( relPermAccessors.get( fields::relperm::dPhaseRelPerm_dPhaseVolFraction {} )),
+    m_wellElemGravCoef( subRegion.getField< fields::well::gravityCoefficient >()),
+    m_wellElemPres( subRegion.getField< fields::well::pressure >()),
+    m_wellElemCompDens( subRegion.getField< fields::well::globalCompDensity >()),
+    m_wellElemTotalMassDens( subRegion.getField< fields::well::totalMassDensity >()),
+    m_dWellElemTotalMassDens( subRegion.getField< fields::well::dTotalMassDensity >()),
+    m_wellElemCompFrac( subRegion.getField< fields::well::globalCompFraction >()),
+    m_dWellElemCompFrac_dCompDens( subRegion.getField< fields::well::dGlobalCompFraction_dGlobalCompDensity >()),
+    m_perfGravCoef( perforationData->getField< fields::well::gravityCoefficient >()),
+    m_perfWellElemIndex( perforationData->getField< fields::perforation::wellElementIndex >()),
+    m_perfTrans( perforationData->getField< fields::perforation::wellTransmissibility >()),
+    m_resElementRegion( perforationData->getField< fields::perforation::reservoirElementRegion >()),
+    m_resElementSubRegion( perforationData->getField< fields::perforation::reservoirElementSubRegion >()),
+    m_resElementIndex( perforationData->getField< fields::perforation::reservoirElementIndex >()),
+    m_compPerfRate( perforationData->getField< fields::well::compPerforationRate >()),
+    m_dCompPerfRate( perforationData->getField< fields::well::dCompPerforationRate >()),
+    m_disableReservoirToWellFlow( disableReservoirToWellFlow )
+  {}
+
+  struct StackVariables
+  {
+public:
+    /**
+     * @brief Constructor for the stack variables
+     */
+
+    GEOS_HOST_DEVICE
+    StackVariables() {}
+
+  };
+
+  template< typename STRUCT = NoOpFunc >
+  GEOS_HOST_DEVICE
+  inline
+  void
+  computeFlux( localIndex const iperf, STRUCT && stack= NoOpFunc{} ) const
```

Comment:
```suggestion
  computeFlux( localIndex const iperf, StackVariables & stack, FUNC && assemblyKernelOp = NoOpFunc{}) const
```

?
But do you really need that callback? In the compositional flux kernels it is used to add additional operations like those needed to add thermal effects. Here, you seem to be doing it with that `if constexpr` (which I think is a good approach) so I don't think you need the callback at all.

Reply:
pushed restructured version